File: demo/process_users.py
import os
import json
import pandas as pd
from multiprocessing import Pool, cpu_count

# Paths
RATINGS_PATH = '../data/ml-20m-psm/ratings.csv'
USER_GROUPS_PATH = 'user_groups.json'

# User groups
user_groups = {"lessthan50": [], "from51to150": [], "from151to300": [], "morethan300": [], "all": []}

# Function to categorize users
def categorize_user(row):
    user_id, count = row
    if count < 50:
        return ("lessthan50", user_id)
    elif 50 <= count <= 150:
        return ("from51to150", user_id)
    elif 150 < count <= 300:
        return ("from151to300", user_id)
    elif count > 300:
        return ("morethan300", user_id)
    return None

# Process ratings to categorize users by number of ratings
def process_users():
    global user_groups

    print("Processing ratings to generate user groups...")
    ratings = pd.read_csv(RATINGS_PATH, sep=',', engine='python', names=['userId', 'movieId', 'rating', 'timestamp'], skiprows=0)

    # Ensure all values are correctly parsed
    ratings['userId'] = pd.to_numeric(ratings['userId'], errors='coerce')
    ratings = ratings.dropna(subset=['userId'])
    ratings['userId'] = ratings['userId'].astype(int)

    # Count ratings per user
    rating_counts = ratings['userId'].value_counts().reset_index()
    rating_counts.columns = ['userId', 'count']

    # Use multiprocessing to categorize users
    with Pool(cpu_count()) as pool:
        results = pool.map(categorize_user, rating_counts.itertuples(index=False, name=None))

    # Aggregate results into user groups
    for result in results:
        if result:
            group, user_id = result
            user_groups[group].append(user_id)
    
    user_groups['all'] = rating_counts['userId'].tolist()

    # Save user groups to file
    with open(USER_GROUPS_PATH, 'w') as f:
        json.dump(user_groups, f)

    print("User groups saved to file.")

if __name__ == '__main__':
    process_users()

File: demo/process_users.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

const RATINGS_PATH = path.join(__dirname, '../data/ml-20m-psm/ratings.csv');
const USER_GROUPS_PATH = path.join(__dirname, 'user_groups.json');

let userGroups = { single: [], ten: [], fifty: [], hundred: [], all: [] };

// Process ratings to categorize users by number of ratings
fs.createReadStream(RATINGS_PATH)
  .pipe(csv({ separator: ',', headers: ['userId', 'movieId', 'rating', 'timestamp'] }))
  .on('data', (row) => {
    if (!userGroups.all.includes(row.userId)) {
      userGroups.all.push(row.userId);
    }
  })
  .on('end', () => {
    const userRatingsCount = {};

    // Calculate ratings count per user
    userGroups.all.forEach((userId) => {
      userRatingsCount[userId] = (userRatingsCount[userId] || 0) + 1;
    });

    // Categorize users based on their ratings count
    Object.entries(userRatingsCount).forEach(([userId, count]) => {
      if (count === 1) userGroups.single.push(userId);
      else if (count >= 5 && count <= 15) userGroups.ten.push(userId);
      else if (count >= 35 && count <= 65) userGroups.fifty.push(userId);
      else if (count >= 60 && count <= 140) userGroups.hundred.push(userId);
    });

    // Persist user groups to file
    fs.writeFileSync(USER_GROUPS_PATH, JSON.stringify(userGroups, null, 2));
    console.log('User groups saved to file:', USER_GROUPS_PATH);
  });
